eth_account/_utils/structured_data/hashing.py

```diff
@@ -0,0 +1,212 @@
+
+
+def is_valid_abi_type(type_name):
+    """
+    This function is used to make sure that the ``type_name`` is a valid ABI Type.
+
+    Please note that this is a temporary function and should be replaced by the corresponding
+    ABI function, once the following issue has been resolved.
+    https://github.com/ethereum/eth-abi/issues/125
+    """
+    valid_abi_types = {"address", "bool", "bytes", "int", "string", "uint"}
+    is_bytesN = type_name.startswith("bytes") and 1 <= int(type_name[5:]) <= 32
+    is_intN = (
+        type_name.startswith("int") and
+        8 <= int(type_name[3:]) <= 256 and
+        int(type_name[3:]) % 8 == 0
+    )
+    is_uintN = (
+        type_name.startswith("uint") and
+        8 <= int(type_name[4:]) <= 256 and
+        int(type_name[4:]) % 8 == 0
+    )
+
+    if type_name in valid_abi_types:
+        return True
+    elif is_bytesN:
+        # bytes1 to bytes32
+        return True
+    elif is_intN:
+        # int8 to int256
+        return True
+    elif is_uintN:
+        # uint8 to uint256
+        return True
+
+    return False
+
+
+@to_tuple
+def _encodeData(primaryType, types, data):
+    # Add typehash
+    yield "bytes32", typeHash(primaryType, types)
+
+    # Add field contents
+    for field in types[primaryType]:
+        value = data[field["name"]]
+        if field["type"] == "string":
+            if not isinstance(value, str):
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "string value".format(
+                        field["name"],
+                        primaryType,
+                        value,
+                        type(value),
+                    )
+                )
+            # Special case where the values need to be keccak hashed before they are encoded
+            hashed_value = keccak(text=value)
+            yield "bytes32", hashed_value
+        elif field["type"] == "bytes":
+            if not isinstance(value, bytes):
+                raise TypeError(
+                    "Value of `{0}` ({2}) in the struct `{1}` is of the type `{3}`, but expected "
+                    "bytes value".format(
+                        field["name"],
+                        primaryType,
+                        value,
+                        type(value),
+                    )
+                )
+            # Special case where the values need to be keccak hashed before they are encoded
+            hashed_value = keccak(primitive=value)
+            yield "bytes32", hashed_value
+        elif field["type"] in types:
+            # This means that this type is a user defined type
+            hashed_value = keccak(primitive=encodeData(field["type"], types, value))
+            yield "bytes32", hashed_value
+        elif field["type"][-1] == "]":
+            # TODO: Replace the above conditionality with Regex for identifying arrays declaration
```

Comment:
cc @davesque 

So we have a very nice mechanism in eth-abi to do this but it isn't a public API.  Maybe it should be.

https://github.com/ethereum/eth-abi/blob/master/eth_abi/grammar.py

This approach is quite robust (more so than a regex based approach) and might be worth exploring how we could expose that functionality for use here.

Reply:
The `eth_abi.grammar` module is public API after some recent changes.

Reply:
https://github.com/ethereum/eth-abi/blob/master/docs/grammar.rst